### ISSUE TITLE
Issue with dates.

### ISSUE THREAD

**Issue by BE-Webdesign** [WP-API/client-js#137]
In reference to https://github.com/WP-API/WP-API/issues/2019. There are strange things that happen with dates and timezones.  I have experienced this myself using wp-api.js

**Comment by nikolov-tmw** [comment]
@BE-Webdesign asked me to post about this here as well.

I think the easiest way to reproduce the issue is to have your computer's timezone(and that of your browser respectively) be different than the timezone of your server(or is it the WordPress install? - I can't say for certain).

That's when weird stuff happens. And I guess it really isn't that weird, because there's just a mismatch of the timezone your client uses to speak to the server. And since neither the client is aware of the server's timezone or vice-versa, you end-up with time differences. Which could be a pretty big deal if you expect something to get published right away and it ends up being scheduled to publish in the future.

I've posted my workaround for the issue in the other thread and can post it here if you'd like. In short I ended up overwriting the base models for each post type that I use the API for and specifically their `toJSON` and `parse` methods. That way I was able to convert the dates to either the client's timezone or the server's timezone, depending on which way the data is going.

**Comment by adamsilverstein** [comment]
@nikolov-tmw thanks for raising this issue. 

I agree this is going to cause some unexpected results for users and needs better support, documentation and testing. One issue I think we have currently is the client has no way of knowing what time zone the server is in. Without this information the client is assuming UTC when it creates a Date object from the returned timestamp. If the API gets updated to enable retrieving the timezone or includes the timezone in the post response, the client could adjust the created Date.

**Comment by nikolov-tmw** [comment]
@adamsilverstein I'm curious how this seems to have slipped past most people(not blaming anyone here, just curious 😃 ).

Just a minor correction though - in my experience the Date object seems to be created in whatever timezone the user's browser(OS) is. For instance on my end when I do this(a random date from an API response):

`new Date( '2014-07-04T01:41:57' )`

I get this in the console:

`Fri Jul 04 2014 04:41:57 GMT+0300 (Eastern Europe Daylight Time)`

**Comment by rmccue** [comment]
Try tacking a `Z` on to the end of the string you're passing into the Date constructor.

We don't specify any timezone as a deliberate decision, since we cannot be authoritative; the JS client needs to handle this itself when parsing.

**Comment by nikolov-tmw** [comment]
Like so: `new Date('2014-07-04T01:41:57Z')` ? Still the same result - `Fri Jul 04 2014 04:41:57 GMT+0300 (Eastern Europe Daylight Time)`(in Chrome's Developer Tools Console).

I can see where you're coming from and I'm personally ok with it. However I think this should be better documented(I don't remember seeing it being mentioned anywhere back when I stumbled across the issue at first) and it should be easier for clients(I guess mostly browsers and less PHP/other languages) to deal with dates.

Actually, reading through the main response of the API `/wp-json`, I see that the `date` field is described as `The date the object was published, in the site's timezone.`. Which technically defines it, but is honestly the first time I've noticed it.

I understand that either writing a bunch of time-zone handling functionality or shipping with a library like moment would be way over the top, so what are your thoughts on including a `timezone` property in the main body of `/wp-json`?
Right now most API consumers are developed(or are they?) for either a specific plugin or a theme, so storing the timezone in a globally accessible object(via `wp_localize_script()` for instance) is no big deal(that is part of my solution). But I can imagine that in the future there will be consumers that don't require you to install anything on your site(think of a post publishing app for instance) - making timezone information accessible would be helpful in such situations.

**Comment by westonruter** [comment]
@nikolov-tmw:

> Like so: `new Date('2014-07-04T01:41:57Z')` ? Still the same result - `Fri Jul 04 2014 04:41:57 GMT+0300 (Eastern Europe Daylight Time)` (in Chrome's Developer Tools Console).

This shows that it is working. Notice how you entered `1:41:57` UTC and got back a time that is 3 hours later at `04:41:57` which corresponds to your local timezone offset at GMT+0300. It's Chrome's dev tools that is outputting a local time, but the `Date` object itself has the expected value.

**Comment by adamsilverstein** [comment]
@nikolov-tmw 

Thanks for raising this issue! I was able to readily reproduce the issue as you described, by setting my computer's timezone to something different than the server. As you mention, loading and saving a post from the API with no changes in this situation changes the stored date on the post. 

I'm still not certain how this happens, when I look at the data sent and returned from the API it looks like I'm sending back the very same data that the API returned which I would expect to work. I'm going to test that as well - receiveing and returning the date strings without modification - to verify that the issue isn't on the API side.

I have two potential solutions so far:

- I developed a patch in that adjust the date parsing to add offsets to the regular times based on the difference between the regular and the gmt times returned from the api: https://github.com/WP-API/client-js/compare/fix-date-timezone?expand=1. This fix is already merged to the develop branch and is ready for testing. In my local testing this resolved the issue, can you please review?

- I am considering dropping date transformations entirely. Date manipulation is pretty inconsistent in browsers and we are adding a lot of complexity here without a clear benefit or use case. If I remove this from the codebase, I can create a shim to allow current users to restore the functionality if they depend on it. How are you using the dates?

One other thing I will look at is how @kadamwhite is handling dates in the https://github.com/WP-API/node-wpapi - perhaps he has already solved this issue there.

**Comment by nikolov-tmw** [comment]
@rmccue 

Actually, I still think there's an issue with how the API works regarding date handling. At the very least it differs than the way that `wp_insert_post()` works. Here's an example(latest version of WordPress, timezone set to UTC+3, the REST API plugin and the client library, with Twenty Fifteen).

In functions.php I added the following bit of code:

```php
if ( isset($_GET['newpost']) ) {
	$date = current_time( 'mysql' );
	$post = array(
		'post_type' => 'post',
		'post_title' => 'New Post - ' . $date,
		'post_content' => 'blah',
		'post_status' => 'publish',
	);

	sleep( 5 );

	wp_insert_post( $post );
}
```

This creates a new post with a title like `New Post – 2016-10-12 19:55:52` and a matching date of `2016-10-12 19:55:52`.

Now, in the API(just running through the console as it's easier) I did this:

```js
var new_post = new wp.api.models.Post( { title: 'API Post - 2016-10-12 19:57:05', date: '2016-10-12 19:57:05', content: 'This is an API post', status: 'publish' } );
new_post.save();
```

This on the other hand created for me a post with the title `API Post - 2016-10-12 19:57:05`, with a status of `Scheduled` and a date of `2016-10-12 22:57:05`. I didn't use a Datetime object, and the raw payload of the request is:

`{"title":"API Post - 2016-10-12 19:57:05","date":"2016-10-12 19:57:05","content":"This is an API post","status":"publish"}`

As you can see - the request was with time of `19:57:05` and the response had the `22:57:05` time.

To me that's a clear sign that there's something wrong on the server side of things as the date gets changed. Even in the API itself it says that the date should be in the site's timezone, but when you pass it in the site's timezone, it ends up changing.

I hope my sample code illustrates my point well enough and is going to be easy to reproduce.

**Comment by adamsilverstein** [comment]
@nikolov-tmw I have a fix for this: https://github.com/WP-API/client-js/pull/140 I'll build in develop if you would like to test that would be great!

**Comment by nikolov-tmw** [comment]
@adamsilverstein I can confirm that this fixes the issue with fetching and saving the post without adjusting the date.

I also see where the root of the problem is coming from in my previous example. The API expects GMT dates, and not dates in the site's timezone. By passing a datetime string in the site's timezone, the JS library can't turn my date into a UTC date when converting to JSON before sending to the server.

@rmccue - can you confirm that this is the case? I don't mind if this continues functioning in the same way(although it might be confusing for people that don't use `Date` objects in JS, since it does things differently than the `wp_insert_post()` counterpart) as long as we make it clear that the date has to be UTC/GMT(so change these [two](https://github.com/WP-API/WP-API/blob/develop/lib/endpoints/class-wp-rest-posts-controller.php#L1469) [notes](https://github.com/WP-API/WP-API/blob/develop/lib/endpoints/class-wp-rest-posts-controller.php#L1514) and the accompanying docs). It would also be a good idea to mention that when you use `Date` objects with the JS client, the conversion will be done automatically.

**Comment by adamsilverstein** [comment]
@nikolov-tmw I think I am going with a different approach you can test in develop or review my changes in this PR: https://github.com/WP-API/client-js/pull/141

basically, I am no longer auto processing date fields when you fetch or save models to the API. Instead, i have introduced two helpers on the models `getDate` and `setDate` that provide the same transformations if you want them (you could still auto-parse them by hooking into toJSON and parse). This way developers can decide how they want to use the date fields. The helpers doc block tries to explain how the dates should be used.

I'm still unsure - this might be a problem? does the API return the date in the server's timezone when getting, but when setting expect the date in UTC?

**Comment by rmccue** [comment]
`date` should definitely be in the site's timezone when you send it to the server, not UTC. If that isn't the case, it's definitely a bug.

**Comment by adamsilverstein** [comment]
@rmccue Worth investigating this a bit further as it may be a bug on the API side. I can try to write some unit tests to check. In my testing retrieving a post and saving it with the client, the date worked if I created the JS Date object from `date_gmt` and sent that back. If I used `date`, the post time would get offset by the difference between the server's timezone and UTC (5 hours in my case).

For now I am going to change the client so it doesn't auto-parse dates, instead offering helper methods to enable date conversions. See https://github.com/WP-API/client-js/pull/141.
